### PR TITLE
[bnxt] Add support for BCM957608

### PR DESCRIPTION
**Pull request by jw14812**
Add support for BCM957608 device.  Add support for additional link speeds supported by BCM957608.

### CODE REVIEW

**Comment by mcb30** [comment]
@jw14812 I've updated the commit with your author name and email, and added the missing `Signed-off-by` line.

Could you please:

1. Run
   ```
   git config --global user.name "Joseph Wong"
   git config --global user.email "joseph.wong@broadcom.com"
   ```
   so that any future PRs will already have the correct user name and email in place
2. Create commits using `git commit -s` in future, so that the `Signed-off-by` line is present
3. Confirm that you are happy with the author changes and the `Signed-off-by` line that I have made to this commit of yours?

Thanks!

File: src/drivers/net/bnxt/bnxt.c [hunk 1]
```diff
@@ -82,16 +82,17 @@ static struct pci_device_id bnxt_nics[] = {
 	PCI_ROM( 0x14e4, 0x1807, "14e4-1807", "14e4-1807", BNXT_FLAG_PCI_VF ),
 	PCI_ROM( 0x14e4, 0x1808, "14e4-1808", "14e4-1808", BNXT_FLAG_PCI_VF ),
 	PCI_ROM( 0x14e4, 0x1809, "14e4-1809", "14e4-1809", BNXT_FLAG_PCI_VF ),
+	PCI_ROM( 0x14e4, 0x1760, "14e4-1760", "14e4-1760", 0 ),
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt.c:85]: The `14e4-17xx` should be **before** `14e4-18xx`.

File: src/drivers/net/bnxt/bnxt.c [hunk 2]
```diff
@@ -82,16 +82,17 @@ static struct pci_device_id bnxt_nics[] = {
...
 };
 
 /**
  * Check if Virtual Function
  */
 u8 bnxt_is_pci_vf ( struct pci_device *pdev )
 {
-	if ( FLAG_TEST ( pdev->id->driver_data, BNXT_FLAG_PCI_VF ) ) {
+	if ( FLAG_TEST ( pdev->id->driver_data, BNXT_FLAG_PCI_VF ) )
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt.c:93]: The removal of the `{` should not be in this commit.

File: src/drivers/net/bnxt/bnxt.c [hunk 3]
```diff
@@ -82,16 +82,17 @@ static struct pci_device_id bnxt_nics[] = {
...
 		return 1;
-	}
+
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt.c:95]: The removal of the `}` should not be in this commit.

File: src/drivers/net/bnxt/bnxt_dbg.h [hunk 4]
```diff
@@ -475,7 +475,7 @@ void dbg_rx_stat(struct bnxt *bp)
 #endif
 
 #if defined(DEBUG_CQ)
-static void dump_cq(struct cmpl_base *cmp, u16 cid)
+static void dump_cq(struct cmpl_base *cmp, u16 cid, u8 toggle)
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt_dbg.h:478]: Neither BCM957608 specific.

File: src/drivers/net/bnxt/bnxt_hsi.h [hunk 5]
```diff
@@ -3249,7 +3316,31 @@ struct hwrm_port_phy_qcfg_output {
 	#define PORT_PHY_QCFG_RESP_PHY_TYPE_200G_BASESR4     0x1dUL
 	#define PORT_PHY_QCFG_RESP_PHY_TYPE_200G_BASELR4     0x1eUL
 	#define PORT_PHY_QCFG_RESP_PHY_TYPE_200G_BASEER4     0x1fUL
-	#define PORT_PHY_QCFG_RESP_PHY_TYPE_LAST            PORT_PHY_QCFG_RESP_PHY_TYPE_200G_BASEER4
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt_hsi.h:3252]: `200`

File: src/drivers/net/bnxt/bnxt_hsi.h [hunk 6]
```diff
@@ -3249,7 +3316,31 @@ struct hwrm_port_phy_qcfg_output {
...
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_200G_BASEER2     0x2fUL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASECR8     0x30UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASESR8     0x31UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASELR8     0x32UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASEER8     0x33UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASECR4     0x34UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASESR4     0x35UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASELR4     0x36UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASEER4     0x37UL
+	#define PORT_PHY_QCFG_RESP_PHY_TYPE_LAST            PORT_PHY_QCFG_RESP_PHY_TYPE_400G_BASEER4
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt_hsi.h:3343]: `400`

File: src/drivers/net/bnxt/bnxt.c (hunk 1 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt.c]: Ack.  I have moved this new entry below the last 14e4-175x entry.  I will reorder the other out of order entries in a different commit

File: src/drivers/net/bnxt/bnxt.c (hunk 2 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt.c]: Ack, Reinserted the brace, and will remove it under a new commit

File: src/drivers/net/bnxt/bnxt.c (hunk 3 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt.c]: Ack.  Reinserted, and will remove under a new commit

File: src/drivers/net/bnxt/bnxt_dbg.h (hunk 4 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt_dbg.h:478]: This new field is specific to BCM57608 functionality

File: src/drivers/net/bnxt/bnxt.c (hunk 2 above)
Comment (stappersg) [reply src/drivers/net/bnxt/bnxt.c]: And that commit will explain **why**.

File: src/drivers/net/bnxt/bnxt_dbg.h (hunk 4 above)
Comment (stappersg) [reply src/drivers/net/bnxt/bnxt_dbg.h:478]: Yeah, my original _"Neither BCM957608 specific."_ should have been something like _"That doesn't look BCM957608 specific, why is it here in this commit?"_.

Comment (jw14812) [reply src/drivers/net/bnxt/bnxt_dbg.h:478]: Understood.  An additional field was added for completion processing. The existing completion processing was updated to handle this new field, rather than creating a separate function.  Would it be helpful if comments were added into the code to explain this?

File: src/drivers/net/bnxt/bnxt_hsi.h (hunk 6 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt_hsi.h:3343]: BCM57608 devices support 400G, thus the higher link speeds values have been included in this commit.

File: src/drivers/net/bnxt/bnxt_dbg.h (hunk 4 above)
Comment (stappersg) [reply src/drivers/net/bnxt/bnxt_dbg.h:478]: The question _"why is it here in this commit?"_ reworded as a request (with **why**) _"Please put the debug extension into a separate commit. Have 1 commit to 1 functional change. Make it possible that `git bisect` can identify the exact breaking change. Express that you understand that you are changing existing code that is in actual use."_

File: src/drivers/net/bnxt/bnxt.c [hunk 7]
```diff
@@ -1201,55 +1299,75 @@ static int bnxt_get_link_speed ( struct bnxt *bp )
...
-	case LINK_SPEED_DRV_2_5G:
+	case LINK_SPEED_FW_2_5G:
 		bp->medium = SET_MEDIUM_SPEED ( bp, MEDIUM_SPEED_2500MBPS );
 		break;
-	case LINK_SPEED_DRV_10G:
+	case LINK_SPEED_FW_10G:
 		bp->medium = SET_MEDIUM_SPEED ( bp, MEDIUM_SPEED_10GBPS );
 		break;
-	case LINK_SPEED_DRV_25G:
+	case LINK_SPEED_FW_25G:
```
Comment (stappersg) [inline src/drivers/net/bnxt/bnxt.c:1337]: Why the change of `LINK_SPEED_DRV_25G` to `LINK_SPEED_FW_25G`?

File: src/drivers/net/bnxt/bnxt_dbg.h (hunk 4 above)
Comment (jw14812) [reply src/drivers/net/bnxt/bnxt_dbg.h:478]: I apologize if the meaning of my last sentence did not convey my proper meaning.  I really would like your input if comments would help to clarify why this field was added

**Comment by jw14812** [comment]
@mcb30 - I believe I have addressed the comments brought up in this thread thus far.  I will submit a new follow-up patch to address the other concerns brought up once this one has been merged. Thanks

**Comment by stappersg** [comment]
On Wed, Feb 07, 2024 at 10:37:49AM -0800, Joseph Wong wrote:
> @mcb30 - I believe I have addressed the comments brought up in this thread thus far.  I will submit a new follow-up patch to address the other concerns brought up once this one has been merged.

#723

**Comment by NiKiZe** [comment]
You will need to rebase to a single commit, as iPXE keeps a clean git history.

**Comment by jw14812** [comment]
> You will need to rebase to a single commit, as iPXE keeps a clean git history.

Ack.

**Comment by mcb30** [comment]
@jw14812 Merged, thank you for the update!
